iree/samples/static_library/static_library_demo.c

```diff
@@ -0,0 +1,210 @@
+  iree_task_executor_release(executor);
+  for (iree_host_size_t i = 0; i < loader_count; ++i) {
+    iree_hal_executable_loader_release(loaders[i]);
+  }
+  return iree_ok_status();
+}
+
+iree_status_t Run() {
+  // TODO(benvanik): move to instance-based registration.
+  IREE_RETURN_IF_ERROR(iree_hal_module_register_types());
+
+  iree_vm_instance_t* instance = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_vm_instance_create(iree_allocator_system(), &instance));
+
+  iree_hal_device_t* device = NULL;
+  IREE_RETURN_IF_ERROR(create_sample_device(&device), "create device");
+  iree_vm_module_t* hal_module = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_hal_module_create(device, iree_allocator_system(), &hal_module));
+
+  // Load bytecode module from the embedded data.
+  const struct iree_file_toc_t* module_file_toc =
+      iree_samples_static_library_module_create();
+
+  iree_vm_module_t* bytecode_module = NULL;
+  iree_const_byte_span_t module_data =
+      iree_make_const_byte_span(module_file_toc->data, module_file_toc->size);
+  IREE_RETURN_IF_ERROR(iree_vm_bytecode_module_create(
+      module_data, iree_allocator_null(), iree_allocator_system(),
+      &bytecode_module));
+
+  // Allocate a context that will hold the module state across invocations.
+  iree_vm_context_t* context = NULL;
+  iree_vm_module_t* modules[] = {hal_module, bytecode_module};
+  IREE_RETURN_IF_ERROR(iree_vm_context_create_with_modules(
+      instance, &modules[0], IREE_ARRAYSIZE(modules), iree_allocator_system(),
+      &context));
+  iree_vm_module_release(hal_module);
+  iree_vm_module_release(bytecode_module);
+
+  // Lookup the entry point function.
+  // Note that we use the synchronous variant which operates on pure type/shape
+  // erased buffers.
+  const char kMainFunctionName[] = "module.simple_mul";
+  iree_vm_function_t main_function;
+  IREE_RETURN_IF_ERROR(iree_vm_context_resolve_function(
+      context, iree_make_cstring_view(kMainFunctionName), &main_function));
+
+  // Allocate buffers that can be mapped on the CPU and that can also be used
+  // on the device. Not all devices support this, but the ones we have now do.
+  const int kElementCount = 4;
+  iree_hal_buffer_t* arg0_buffer = NULL;
+  iree_hal_buffer_t* arg1_buffer = NULL;
+  iree_hal_memory_type_t input_memory_type =
+      IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE;
+  IREE_RETURN_IF_ERROR(iree_hal_allocator_allocate_buffer(
+      iree_hal_device_allocator(device), input_memory_type,
+      IREE_HAL_BUFFER_USAGE_ALL, sizeof(float) * kElementCount, &arg0_buffer));
+  IREE_RETURN_IF_ERROR(iree_hal_allocator_allocate_buffer(
+      iree_hal_device_allocator(device), input_memory_type,
+      IREE_HAL_BUFFER_USAGE_ALL, sizeof(float) * kElementCount, &arg1_buffer));
+
+  // Populate initial values for 4 * 2 = 8.
+  const float kFloat4 = 4.0f;
+  const float kFloat2 = 2.0f;
+  IREE_RETURN_IF_ERROR(iree_hal_buffer_fill(arg0_buffer, 0, IREE_WHOLE_BUFFER,
```

Comment:
we should switch this sample to the runtime API - scott has examples that should be able to be copy/pasted into this for how to setup inputs/etc - it's a lot less code.

Reply:
https://github.com/google/iree/blob/main/iree/runtime/demo/hello_world_explained.c
https://github.com/google/iree/blob/main/iree/runtime/demo/hello_world_terse.c
https://github.com/google/iree/blob/main/iree/samples/variables_and_state/main.c

I'd recommend referencing/forking the recently added variables_and_state sample, but the others are good too (just a slightly different way of splitting into helper functions)

Reply:
also use `iree_hal_buffer_view_clone_heap_buffer` (or wrap, if it works) instead of `iree_hal_buffer_fill` etc.

https://github.com/google/iree/blob/79cf15440a6be860a8821e5cb520d62a8829a216/iree/hal/buffer_view.h#L137-L146